src/handlers/ecommerce/auth-request.ts

```diff
@@ -1,16 +1,112 @@
+import * as TE from "fp-ts/lib/TaskEither";
+import { v4 as uuid } from "uuid";
+import * as O from "fp-ts/Option";
+import * as E from "fp-ts/lib/Either";
+import { pipe } from "fp-ts/lib/function";
+import { FlowCase, getFlowCookie } from "../../flow";
 import {
   createSuccessAuthRequestResponseEntity,
+  createSuccessAuthRequestResponseEntityFromNPG,
+  error400InvalidRequestBody,
   error404TransactionIdNotFound,
   error409TransactionAlreadyProcessed
 } from "../../utils/ecommerce/auth-request";
-import { FlowCase, getFlowCookie } from "../../flow";
+import { logger } from "../../logger";
+import { RequestAuthorizationRequest } from "../../generated/ecommerce/RequestAuthorizationRequest";
+import {
+  PaymentInstrumentDetail2,
+  PaymentInstrumentDetail3
+} from "../../generated/ecommerce/PaymentInstrumentDetail";
+import { RequestAuthorizationResponse } from "../../generated/ecommerce/RequestAuthorizationResponse";
+
+const PSP_API_KEY = process.env.PSP_API_KEY;
+
+const confirmPaymentFromNpg = async (
+  _req: any,
+  res: any,
+  requestSessionId: string
+): Promise<void> => {
+  const postData = JSON.stringify({
+    amount: _req.body.amount,
+    sessionId: requestSessionId
+  });
 
-const returnSuccessResponse = (res: any): void => {
-  logger.info("[Auth-request ecommerce] - Return success case");
-  res.status(200).send(createSuccessAuthRequestResponseEntity());
+  logger.info(
+    `[Invoke NPG confirm payment with npg-session id: ${requestSessionId}] - Return success case`
+  );
+  const correlationId = uuid();
+  const url = `https://stg-ta.nexigroup.com/api/phoenix-0.0/psp/api/v1/build/confirm_payment`;
+  const response = await fetch(url, {
+    body: postData,
+    headers: {
+      "Content-Type": "application/json",
+      "Correlation-Id": correlationId,
+      "X-Api-key": PSP_API_KEY as string
+    },
+    method: "POST"
+  });
+
+  await pipe(
+    TE.tryCatch(
+      async () => response.json(),
+      _e => {
+        logger.error("Error invoking npg order build");
+      }
+    ),
+    TE.map(resp => {
+      pipe(
+        resp,
+        createSuccessAuthRequestResponseEntityFromNPG,
+        RequestAuthorizationResponse.decode,
+        E.fold(
+          () => {
+            logger.error("Error while invoke NPG unexpected body");
+            res.status(response.status).send(resp);
+          },
+          responseBody => res.status(response.status).send(responseBody)
+        )
+      );
+    }),
+    TE.mapLeft(() => res.status(response.status).send())
+  )();
+};
+
+const processAuthorizationRequest = (req: any, res: any): void => {
+  pipe(
+    req.body,
+    RequestAuthorizationRequest.decode,
+    O.fromEither,
+    O.fold(
+      () => res.status(409).send(error400InvalidRequestBody()),
+      resp => {
+        pipe(
+          resp.details,
+          PaymentInstrumentDetail3.decode,
+          O.fromEither,
+          O.fold(
+            () => {
+              pipe(
+                resp.details,
```

Comment:
Given that you use the same error handler (i.e. 409 status + error body) you can refactor this using multiple `O.map`/`O.chain` and then use a single `O.fold` at the end. This way you avoid nesting too many callbacks

Reply:
Fixed with 67b8b2188d6338647d5d5b7135ab2734a5fdc949